src/briefcase/integrations/subprocess.py

```diff
@@ -206,6 +210,39 @@ def Popen(self, args, **kwargs):
             **self.final_kwargs(**kwargs)
         )
 
+    def stream_output(self, popen_process):
+        """
+        Stream the output of a Popen process until the process exits.
+        If the user sends CTRL+C, the process will be terminated.
+
+        This is useful for starting a process via Popen such as tailing a
+        log file, then initiating a non-blocking process that populates that
+        log, and finally streaming the original process's output here.
+
+        :param popen_process: a running Popen process with output to print
+        """
+        try:
+            while True:
+                # readline should always return at least a newline...if it returns an
+                # empty string, that should mean the underlying process is exiting/gone
+                output_line = ensure_str(popen_process.stdout.readline())
+                return_code = popen_process.poll()
+                if output_line == "" and return_code is not None:
```

Comment:
I was a little leery about this loop as well at first. However, I believe it hinges on assumptions that are already baked in to our other uses of `subprocess`.

- `check_output` is wrapper on `run` and `run` is a wrapper on `Popen`.
- `Popen` does some basic validation, sets up `stdin/out/err`, and then forks a process to run the OS command.
  - Strictly speaking, `Popen` can fail but it appears to be either invalid input combinations or something went really wrong at the OS level.
- From here, `run` blocks on `read\write` calls to `stdin/out/err` via `Popen.communicate()`.
  - These `read` call implementations are buried in layers of abstraction, but all the documentation of python `io` suggests that `read` returns when it hits an `EOF`.
  - For `readline`, the docs say that once an `EOF` is hit, `readline` will only ever return `""` or `b""` for subsequent calls.
  - For `stdout/err`, I am interpreting `EOF` to mean that the child process closed the pipes on its end to them.
- Once `Popen.communicate()` returns, `run` blocks on `wait()`.
  - True to its name, it waits until the child process finishes and provides a return code.
- If return code is non-zero, raise `CalledProcessError`
- Alternatively, return a `CompletedProcess` object with return code and `stdout/err`

So....in this way, this loop is a really dumbed down version of `run`, in that:
- At some point, `readline()` will need to start returning `""` because the process closed the pipe
- And the child process will need to return an code via `poll()`

If these assumptions aren't met from a child process, then we'll need the user to send CTRL+C.....however, that would be true even using `run` (at least without a `timeout` since we don't provide one).


As for the specific structure of the loop:
 - The quit condition is based on both `readline()` returning empty and `poll()` providing a return code.
 - This is mostly to prevent a race condition with flushing `readline` output to the screen and `poll()` getting a return code from the process.
 - Requiring both means we've flushed all the output and we have a return code.
 - That said, I think there are other configurations for the loop, but I don't think they fundamentally change the need for `readline` returns `""` and `poll()` is not `None`.
 - Finally, this `Popen` loop idiom was surprisingly popular when I was researching the best way to do this....although, it's always hard to tell if that's because everyone just copy/pastes the seemingly most popular thing...or if its actually the best way to do this.

Reply:
Ok - this all makes sense, and the revised loop structure looks good to me. `poll()` is definitely less of a risk than `wait()` under the circumstances; and AFAICT the usage of `poll()` you've got here should be safe.